controllers/application_controller.go

```diff
@@ -0,0 +1,247 @@
+// +kubebuilder:rbac:groups=app.k8s.io,resources=applications/status,verbs=get;update;patch
+
+func (r *ApplicationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	rootCtx := context.Background()
+	logger := r.Log.WithValues("application", req.NamespacedName)
+	ctx := context.WithValue(rootCtx, loggerCtxKey, logger)
+
+	var app appv1beta1.Application
+	err := r.Get(ctx, req.NamespacedName, &app)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Application is in the process of being deleted, so no need to do anything.
+	if app.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus := r.getNewApplicationStatus(ctx, &app)
+
+	if equality.Semantic.DeepEqual(newApplicationStatus, &app.Status) {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus.ObservedGeneration = app.Generation
+	if err = r.updateApplicationStatus(ctx, req.NamespacedName, newApplicationStatus); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ApplicationReconciler) updateComponents(ctx context.Context, app *appv1beta1.Application) ([]*unstructured.Unstructured, error) {
+	resources, err := r.fetchComponentListResources(ctx, app.Spec.ComponentGroupKinds, app.Spec.Selector, app.Namespace)
+	if err != nil {
+		return resources, err
+	}
+
+	if app.Spec.AddOwnerRef {
+		ownerRef := metav1.NewControllerRef(app, appv1beta1.GroupVersion.WithKind("Application"))
+		*ownerRef.Controller = false
+		if err := r.setOwnerRefForResources(ctx, *ownerRef, resources); err != nil {
+			return resources, err
+		}
+	}
+	return resources, nil
+}
+
+func (r *ApplicationReconciler) getNewApplicationStatus(ctx context.Context, app *appv1beta1.Application) *appv1beta1.ApplicationStatus {
+
+	resources, err := r.updateComponents(ctx, app)
+
+	objectStatuses := r.objectStatuses(ctx, resources)
+	aggReady := aggregateReady(objectStatuses)
+
+	newApplicationStatus := app.Status.DeepCopy()
+
+	newApplicationStatus.ComponentList = appv1beta1.ComponentList{
+		Objects: objectStatuses,
+	}
+
+	if aggReady {
+		setReadyCondition(newApplicationStatus, "ComponentsReady", "all components ready")
+	} else {
+		setNotReadyCondition(newApplicationStatus, "ComponentsNotReady", "some components not ready")
+	}
+
+	if err != nil {
+		setErrorCondition(newApplicationStatus, "ErrorSeen", err.Error())
+	} else {
+		clearErrorCondition(newApplicationStatus)
+	}
+
+	return newApplicationStatus
+}
+
+func (r *ApplicationReconciler) fetchComponentListResources(ctx context.Context, groupKinds []metav1.GroupKind, selector *metav1.LabelSelector, namespace string) ([]*unstructured.Unstructured, error) {
+	logger := getLoggerOrDie(ctx)
+	var resources []*unstructured.Unstructured
+	for _, gk := range groupKinds {
+		mapping, err := r.Mapper.RESTMapping(schema.GroupKind{
```

Comment:
In this code, we should make sure we can handle situations where the Group in the list of GroupKinds from the App CR actually contains the GroupVersion. According to the spec it should only contain the group, but I know there are places where the group is included and I think this is handled by the old version of the controller.

Reply:
DONE with the `StripVersion` function. Tests added for such case: https://github.com/kubernetes-sigs/application/blob/1abaa21daac7a901f17ce2049c1270cd5ebfe18a/controllers/application_controller_test.go#L171

Reply:
Right StripVersion was added to address https://github.com/kubernetes-sigs/application/issues/102